File: 0440-k-th-smallest-in-lexicographical-order/0440-k-th-smallest-in-lexicographical-order.py
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        cur = 1
        k -= 1
        while k > 0:
            steps = self.getSteps(n, cur, cur+1)
            if steps <= k:
                cur += 1
                k -= steps
            else:
                cur *= 10
                k -= 1
                
        return cur

    def getSteps(self, n: int, n1: int, n2: int) -> int:
        steps = 0
        while n1 <= n:
            steps += min(n+1, n2) - n1
            n1 *= 10
            n2 *= 10
        return steps

File: 0440-k-th-smallest-in-lexicographical-order/0440-k-th-smallest-in-lexicographical-order.java
class Solution 
{
    public int findKthNumber(int n, int k) 
    {
        int curr = 1; 
        k -= 1; 
        while (k > 0) 
        {
            long steps = countSteps(n, curr, curr + 1); 

            if (steps <= k) 
            {
                curr += 1;   
                k -= steps;
            } 
            else 
            {
                curr *= 10;  
                k -= 1;
            }
        }

        return curr; 
    }

    private long countSteps(int n, long curr, long next) 
    {
        long steps = 0;
        while (curr <= n) 
        {
            steps += Math.min(n + 1L, next) - curr;
            curr *= 10;
            next *= 10;
        }
        
        return steps;
    }
}